src/python/pants/core/goals/tailor.py

```diff
@@ -0,0 +1,382 @@
+        owned_sources: Iterable[str],
+        *,
+        kwargs: Mapping[str, str | int | bool | Tuple[str, ...]] | None = None,
+        comments: Iterable[str] = tuple(),
+        build_file_name: str = "BUILD",
+    ) -> None:
+        self.path = path
+        self.name = name
+        self.type_alias = type_alias
+        self.triggering_sources = tuple(triggering_sources)
+        self.owned_sources = tuple(owned_sources)
+        self.kwargs = FrozenDict(kwargs or {})
+        self.comments = tuple(comments)
+        self.build_file_name = build_file_name
+
+    @property
+    def build_file_path(self) -> str:
+        return os.path.join(self.path, self.build_file_name)
+
+    @property
+    def address(self) -> Address:
+        return Address(self.path, target_name=self.name)
+
+    def rename(self, new_name: str) -> PutativeTarget:
+        """A copy of this object with the name replaced to the given name."""
+        # We assume that a rename imposes an explicit "name=" kwarg, overriding any previous
+        # explicit "name=" kwarg, even if the rename happens to be to the default name.
+        return dataclasses.replace(self, name=new_name, kwargs={**self.kwargs, "name": new_name})
+
+    def restrict_sources(self) -> PutativeTarget:
+        """A copy of this object with the sources explicitly set to just the triggering sources."""
+        owned_sources = self.triggering_sources
+        return dataclasses.replace(
+            self,
+            owned_sources=owned_sources,
+            kwargs={**self.kwargs, "sources": owned_sources},
+        )
+
+    def add_comments(self, comments: Iterable[str]) -> PutativeTarget:
+        return dataclasses.replace(self, comments=self.comments + tuple(comments))
+
+    def generate_build_file_stanza(self, indent: str) -> str:
+        def fmt_val(v) -> str:
+            if isinstance(v, str):
+                return f'"{v}"'
+            if isinstance(v, tuple):
+                val_parts = [f"\n{indent*2}{fmt_val(x)}" for x in v]
+                val_str = ",".join(val_parts)
+                return f"[{val_str},\n{indent}]"
+            return repr(v)
+
+        if self.kwargs:
+            kwargs_str_parts = [f"\n{indent}{k}={fmt_val(v)}" for k, v in self.kwargs.items()]
+            kwargs_str = ",".join(kwargs_str_parts) + ",\n"
+        else:
+            kwargs_str = ""
+
+        comment_str = ("\n".join(self.comments) + "\n") if self.comments else ""
+        return f"{comment_str}{self.type_alias}({kwargs_str})\n"
+
+
+class PutativeTargets(DeduplicatedCollection[PutativeTarget]):
+    sort_input = True
+
+    @classmethod
+    def merge(cls, tgts_iters: Iterable[PutativeTargets]) -> PutativeTargets:
+        all_tgts: List[PutativeTarget] = []
+        for tgts in tgts_iters:
+            all_tgts.extend(tgts)
+        return cls(all_tgts)
+
+
+class TailorSubsystem(GoalSubsystem):
+    name = "tailor"
+    help = "Generate config for working with Pants."
```

Comment:
Would still be good to expand this: at the very least, it would be good to mention that it will always be idempotent, so that folks who already have config don't ignore it.

Reply:
How about `Auto-generate BUILD file targets for new source files.`.